crates/sound/src/device.rs

```diff
@@ -172,9 +178,97 @@ impl VhostUserSoundThread {
 
             let code = ControlMessageKind::try_from(request.code).map_err(Error::from)?;
             match code {
-                ControlMessageKind::ChmapInfo
-                | ControlMessageKind::JackInfo
-                | ControlMessageKind::JackRemap => {
+                ControlMessageKind::ChmapInfo => {
+                    if descriptors.len() != 3 {
+                        log::error!("a CHMAP_INFO request should have three descriptors total.");
+                        return Err(Error::UnexpectedDescriptorCount(descriptors.len()).into());
+                    } else if !descriptors[2].is_write_only() {
+                        log::error!(
+                            "a CHMAP_INFO request should have a writeable descriptor for the info \
+                             payload response after the header status response"
+                        );
+                        return Err(Error::UnexpectedReadableDescriptor(2).into());
+                    }
+                    let request = desc_chain
+                        .memory()
+                        .read_obj::<VirtioSoundQueryInfo>(desc_request.addr())
+                        .map_err(|_| Error::DescriptorReadFailed)?;
+                    let start_id = u32::from(request.start_id) as usize;
+                    let count = u32::from(request.count) as usize;
+                    let chmaps = self.chmaps.read().unwrap();
+                    if chmaps.len() <= start_id || chmaps.len() < start_id + count {
+                        resp.code = VIRTIO_SND_S_BAD_MSG.into();
+                    } else {
+                        let desc_response = descriptors[2];
+                        let mut buf = vec![];
+                        let mut c: VirtioSoundChmapInfo;
+
+                        for i in chmaps
+                            .iter()
+                            .skip(u32::from(request.start_id) as usize)
+                            .take(u32::from(request.count) as usize)
+                        {
+                            c = VirtioSoundChmapInfo::default();
```

Comment:
I would suggest avoiding to use `default()`, but just initialize it with the fields:

```diff
diff --git a/crates/sound/src/device.rs b/crates/sound/src/device.rs
index 2d81a98..0f0bc0b 100644
--- a/crates/sound/src/device.rs
+++ b/crates/sound/src/device.rs
@@ -201,18 +201,20 @@ impl VhostUserSoundThread {
                     } else {
                         let desc_response = descriptors[2];
                         let mut buf = vec![];
-                        let mut c: VirtioSoundChmapInfo;
 
                         for i in chmaps
                             .iter()
                             .skip(u32::from(request.start_id) as usize)
                             .take(u32::from(request.count) as usize)
                         {
-                            c = VirtioSoundChmapInfo::default();
-                            c.hdr.hda_fn_nid = 0.into();
-                            c.direction = i.direction;
-                            c.channels = i.channels;
-                            c.positions = i.positions;
+                            let c = VirtioSoundChmapInfo {
+                                hdr: VirtioSoundInfo {
+                                    hda_fn_nid: 0.into(),
+                                },
+                                direction: i.direction,
+                                channels: i.channels,
+                                positions: i.positions,
+                            };
                             buf.extend_from_slice(c.as_slice());
                         }
                         desc_chain
```

Reply:
But why not using directly `i.as_slice()` instead of creating a new `VirtioSoundChmapInfo`?

Reply:
yeah, right.